vm/odroid/odroid.go

```diff
@@ -0,0 +1,353 @@
+	if inst.cfg.Debug {
+		Logf(0, "running command: ssh %#v", args)
+	}
+	cmd := exec.Command("ssh", args...)
+	cmd.Stdout = wpipe
+	cmd.Stderr = wpipe
+	if err := cmd.Start(); err != nil {
+		wpipe.Close()
+		return nil, err
+	}
+	wpipe.Close()
+
+	done := make(chan bool)
+	go func() {
+		select {
+		case <-time.After(time.Minute):
+			if inst.cfg.Debug {
+				Logf(0, "ssh hanged")
+			}
+			cmd.Process.Kill()
+		case <-done:
+		}
+	}()
+	if err := cmd.Wait(); err != nil {
+		close(done)
+		out, _ := ioutil.ReadAll(rpipe)
+		if inst.cfg.Debug {
+			Logf(0, "ssh failed: %v\n%s", err, out)
+		}
+		return nil, fmt.Errorf("ssh %+v failed: %v\n%s", args, err, out)
+	}
+	close(done)
+	if inst.cfg.Debug {
+		Logf(0, "ssh returned")
+	}
+	out, _ := ioutil.ReadAll(rpipe)
+	return out, nil
+}
+
+func switchPortPower(busNum, deviceNum, portNum int, power bool) error {
+	var context *C.libusb_context
+	if err := C.libusb_init(&context); err != 0 {
+		return fmt.Errorf("failed to init libusb: %v\n", err)
+	}
+	defer C.libusb_exit(context)
+
+	var rawList **C.libusb_device
+	numDevices := int(C.libusb_get_device_list(context, &rawList))
+	if numDevices < 0 {
+		return fmt.Errorf("failed to init libusb: %v", numDevices)
+	}
+	defer C.libusb_free_device_list(rawList, 1)
+
+	var deviceList []*C.libusb_device
+	*(*reflect.SliceHeader)(unsafe.Pointer(&deviceList)) = reflect.SliceHeader{
+		Data: uintptr(unsafe.Pointer(rawList)),
+		Len:  numDevices,
+		Cap:  numDevices,
+	}
+
+	var hub *C.libusb_device
+	for i := 0; i < numDevices; i++ {
+		var desc C.struct_libusb_device_descriptor
+		if err := C.libusb_get_device_descriptor(deviceList[i], &desc); err != 0 {
+			return fmt.Errorf("failed to get device descriptor: %v", err)
+		}
+		if desc.bDeviceClass != C.USB_CLASS_HUB {
+			continue
+		}
+		if C.libusb_get_bus_number(deviceList[i]) != C.uint8_t(busNum) {
+			continue
+		}
+		if C.libusb_get_device_address(deviceList[i]) != C.uint8_t(deviceNum) {
+			continue
+		}
+		hub = deviceList[i]
+		break
+	}
+
+	if hub == nil {
+		return fmt.Errorf("hub not found: bus: %v, device: %v", busNum, deviceNum)
+	}
+
+	var handle *C.libusb_device_handle
+	if err := C.libusb_open(hub, &handle); err != 0 {
+		return fmt.Errorf("failed to open usb device: %v", err)
+	}
+
+	request := C.uint8_t(C.USB_REQ_CLEAR_FEATURE)
+	if power {
+		request = C.uint8_t(C.USB_REQ_SET_FEATURE)
+	}
+	port := C.uint16_t(portNum)
+	timeout := C.uint(1000)
+	if err := C.libusb_control_transfer(handle, C.USB_RT_PORT, request,
+		C.USB_PORT_FEAT_POWER, port, nil, 0, timeout); err < 0 {
+		return fmt.Errorf("failed to send control message: %v\n", err)
+	}
+
+	return nil
+}
+
+func (inst *instance) repair() error {
+	// Try to shutdown gracefully.
+	Logf(1, "odroid: trying to ssh")
+	if err := inst.waitForSsh(10); err == nil {
+		Logf(1, "odroid: ssh succeeded, shutting down now")
+		inst.ssh("shutdown now")
+		if !vm.SleepInterruptible(20 * time.Second) {
```

Comment:
I issue a `shutdown now` command, Odroid starts to reboot, I wait 20 seconds for it to shut down, and then I reset power.